zen_document_parser/base.py

```diff
@@ -0,0 +1,231 @@
+from enum import IntEnum
+from pdfquery import PDFQuery
+from pdfminer.pdfdocument import PDFSyntaxError
+
+from zen_document_parser.exceptions import InvalidPDFError, UnknownVariantError
+
+
+class DocFieldType(IntEnum):
+    TEXT = 1
+    NUMBER = 2
+    CUSTOM = 3  # TODO: Forget this and have 'type' take a callable instead?
+
+
+class DocField(object):
+    def __init__(self, bbox, type=DocFieldType.TEXT, required=False,
+                 description=None):
+        self.bbox = bbox
+        self.type = type
+        self.required = required
+        self.description = description
+
+
+class DocVariantMeta(type):
```

Comment:
The metaclass approach allows for a pattern that people are familiar with (Django, SQLAlchemy, numerous other libs).
Particularly, it allows for a smoother way of overriding the `check_for_match()` method in DocVariant subclasses.
For custom testing logic, instead of typing something like 

```python
if (self.data['form_title'] == SOME_CONST) and (self.data['assessment_year'] == 2014):
    pass
```

the user may feel it is easier to type the following (which other libs allow for):

```python
if (self.form_title == SOME_CONST) and (self.assessment_year == 2014):
    pass
```

It's not an extremely massive saving of keystrokes, but it's still makes things easier, and based on all the libs out there, a lot of people may find this as 'expected behaviour' for a lib that allows specifying 'models' in the same way as Django/SQA/etc.